client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -916,9 +991,99 @@ impl Notifications {
 
 			// Any state other than `Incoming` is invalid.
 			peer => {
-				error!(target: "sub-libp2p",
-					"State mismatch in libp2p: Expected alive incoming. Got {:?}.",
-					peer);
+				error!(
+					target: LOG_TARGET,
+					"State mismatch in libp2p: Expected alive incoming. Got {:?}.", peer
+				);
+				debug_assert!(false);
+			},
+		}
+	}
+
+	/// Protocol rejected inbound substream.
+	fn protocol_report_reject(&mut self, index: sc_peerset::IncomingIndex) {
+		self.peerset_report_reject(index);
+	}
+
+	/// Function that is called when the peerset wants us to accept a connection
+	/// request from a peer.
+	fn peerset_report_accept(&mut self, index: sc_peerset::IncomingIndex) {
+		let incoming = if let Some(pos) = self.incoming.iter().position(|i| i.incoming_id == index)
+		{
+			self.incoming.remove(pos)
+		} else {
+			error!(target: LOG_TARGET, "PSM => Accept({:?}): Invalid index", index);
+			return
+		};
+
+		if !incoming.alive {
+			trace!(
+				target: LOG_TARGET,
+				"PSM => Accept({:?}, {}, {:?}): Obsolete incoming",
+				index,
+				incoming.peer_id,
+				incoming.set_id
+			);
+			match self.peers.get_mut(&(incoming.peer_id, incoming.set_id)) {
+				Some(PeerState::DisabledPendingEnable { .. }) | Some(PeerState::Enabled { .. }) => {
+				},
+				_ => {
+					trace!(
+						target: LOG_TARGET,
+						"PSM <= Dropped({}, {:?})",
+						incoming.peer_id,
+						incoming.set_id
+					);
+					self.peerset.dropped(incoming.set_id, incoming.peer_id, DropReason::Unknown);
+				},
+			}
+			return
+		}
+
+		let state = match self.peers.get_mut(&(incoming.peer_id, incoming.set_id)) {
+			Some(s) => s,
+			None => {
+				debug_assert!(false);
+				return
+			},
+		};
+
+		match mem::replace(state, PeerState::Poisoned) {
+			// Incoming => Enabled
+			PeerState::Incoming { ref mut accepted, backoff_until, connections } => {
+				trace!(
+					target: LOG_TARGET,
+					"PSM => Accept({:?}, {}, {:?}): Send substream for validation to protocol.",
+					index,
+					incoming.peer_id,
+					incoming.set_id
+				);
+
+				match self.protocol_handles[usize::from(incoming.set_id)]
+					.report_incoming_substream(incoming.peer_id, incoming.handshake.clone())
+				{
+					Ok(rx) => {
+						*accepted = true;
```

Comment:
I have been working on several different branches doing the integration and I ultimately ditched this `accepted` approach as too ugly but forgot to update this branch. Basically the current (unpublished) idea is that when `Peerset` has accepted the peer, instead of calling `peerset_report_accept()`, `Notifications` calls `report_incoming_substream()`. If the response for that call is `Accept`, `Notifications` calls `peerset_report_accept()` and the code works as it always has. If the received response is `Reject`, `Notifications` calls `Peerset::rejected()` to release the allocated slot. I'm not too happy with this approach either because ultimately there should only one exchange between `Notifications` and the protocol but I believe it suffices for now.

Reply:
I published the changes. What is left unimplemented is releasing the slot from `Peerset` if the substream was rejected by the protocol. Since I know you have pending changes to the `Peerset` so I'll just save myself the trouble of resolving conflicts and added a TODO instead as a reminder.